1.5.0 release post (#304)
crystal-lang/crystal-website: _releases/2022-07-06-1.5.0-released.md

Pre-built packages are available on [GitHub Releases](https://github.com/crystal-lang/crystal/releases/tag/1.5.0) and our official distribution channels.
See [crystal-lang.org/install](https://crystal-lang.org/install/) for installation instructions.

## Stats

In this release we included [XXX changes since the 1.4.1 release](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+milestone%3A1.5.0) by XX contributors. We thank all the effort put into improving the language! ❤️

## Remarkable changes

### Method restrictions from instance variables

When an instance variable is assigned the value of an untyped method argument, then the argument is restricted to share the same type as the instance variable.

For instance, consider this code:

```crystal
class Foo
  @x : Int64

  def initialize(x)
    @x = x
  end
end
````

Up to 1.4.1, `x` in `initialize` is unrestricted. This had several issues:

>  1. If the user passes an incorrect argument, like `Foo.new('a')`, then the error wrongly blames `@x` for not having the right type.

Reviewer: I wouldn't say this is "wrong" - it's just part of the story. I think we could mention that the reason for error was hard to understand and the new feature makes it easier to work with instance var assignments.

Author: You're correct. I was misunderstanding the message, which is anyway plain wrong:

> Error: instance variable '@x' of Foo must be Int32, not Char

 2. No autocast is performed, for instance, if we pass an `Int32` instead: `Foo.new(1)` fails.

 3. The generated documentation doesn't provide a hint of the type of the parameter `x`.

From 1.5.0, in an assignment like `@x = x`, parameter `x` gets the type of `@x`, effectively solving the three issues above. Details can be read from [#12103](https://github.com/crystal-lang/crystal/pull/12103).

> > Note: if for some reason this new feature is bringing you a headache, let us know in the [issue tracker](https://github.com/crystal-lang/crystal/issues), and disable it passing `-Dno_restrictions_augmenter` in the build options.

Reviewer: ```suggestion
> Note: This new feature is expected to be backwards compatible. If for some reason it brings you a headache you can disable it passing `-Dno_restrictions_augmenter` in the build options. Let us know in the [issue tracker](https://github.com/crystal-lang/crystal/issues) about it.
```

### Annotations allowed on method arguments

It is now possible to add an annotation to a parameter of a method or macro, useful when needing to handle transformations in parameters. For instance, the following fictitious code allows to multiply the value received by the entrypoint `/multiply/{num}`, in this case doubling `num`:

```crystal
@[ARTA::Get(path: "/multiply/{num}")]
def multiply(@[Multiply(by: 2)] num : Int32) : Int32
  num

> end

Reviewer: I don't think this is really a good example. It's unnecessarily complex for illustrating the feature. Alone the fact that there are actually two annotations makes it non-trivial to grok. And it's by far not self-contained. In order to understand what  the code really means, you need some context about the Athena router.

Instead we could perhaps use the example from https://github.com/crystal-lang/crystal/issues/12039#issuecomment-1118859028:
```cr
# This annotation designates that a method parameter could be unused.
# It may be used to suppress a respective linter message.
annotation MaybeUnused; end

def foo(@[MaybeUnused] x)
end 
```
This is still made up and not currently implemented in Ameba or anything, but I think it's better than the router example.

I'm happy for an even better, minimal example.

### Annotations allowed on method arguments

It is now possible to add an annotation to a parameter of a method or macro. As illustration, imagine a linter that warns if a parameter is not used.

```crystal
def foo(x); end  # Warning: argument `x` is not used
```

> Then, we could signal the linter to do not warn us in a particular case. Assume the following annotation provided by the linter:

Reviewer: ```suggestion
Then, we could signal the linter to not warn us in a particular case. Assume the following annotation provided by the linter:
```

Reviewer: ping @beta-ziliani ☝️

```crystal
annotation MaybeUnused; end
```

> Applying it to the parameters removes the warning (in this particular fictitious linter):

Reviewer: ```suggestion
Applying it to the parameter removes the warning (in this particular fictitious linter):
```

puts foo[KEY].size
```

### A big step towards native Windows support

> By merging [#12149](https://github.com/crystal-lang/crystal/pull/12149) we crossed an important check in the [road to native Windows support](https://github.com/crystal-lang/crystal/issues/5430). Additionally, Windows now have a `Makefile`([#11773](https://github.com/crystal-lang/crystal/pull/11773)).

Reviewer: ```suggestion
The concurrency runtime on Windows is now backed by a functioning event loop ([#12149](https://github.com/crystal-lang/crystal/pull/12149)).
This crossed an important check in the [road to native Windows support](https://github.com/crystal-lang/crystal/issues/5430). Additionally, we now have a Windows-compatibly `Makefile`([#11773](https://github.com/crystal-lang/crystal/pull/11773)).
```

### Improvements in the interpreter

> The interpreter is coming to shape! While it won't be released in the binaries distributions yet, we will work towards incorporating it in the major platforms right after this release. Stay tuned!

Reviewer: ```suggestion
The interpreter is coming to shape! [A whole bunch of interpreter bugs](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+sort%3Aupdated-desc+milestone%3A1.5.0+label%3Atopic%3Acompiler%3Ainterpreter) got fixed.
While it won't be released in the binaries distributions yet, we will work towards incorporating it in the major platforms right after this release. Stay tuned!
You can build the compiler with interpreter support locally with `make interpreter=1`, see the [interpreter introduction post for more info](https://crystal-lang.org/2021/12/29/crystal-i.html).
```
